Source/SwiftLintBuiltInRules/Rules/Idiomatic/FileNameRule.swift

```diff
@@ -56,33 +58,118 @@ struct FileNameRule: OptInRule, SourceKitFreeRule {
 }
 
 private class TypeNameCollectingVisitor: SyntaxVisitor {
+    /// All of a visited node's ancestor type names if that node is nested, starting with the furthest
+    /// ancestor and ending with the direct parent
+    private var ancestorNames = Stack<String>()
+
+    /// All of the type names found in the file
     private(set) var names: Set<String> = []
 
+    /// If true, nested types are only allowed in the file name when used by their fully-qualified name
+    /// (e.g. `My.Nested.Type` and not just `Type`)
+    private let requireFullyQualifiedNames: Bool
+
+    init(requireFullyQualifiedNames: Bool) {
+        self.requireFullyQualifiedNames = requireFullyQualifiedNames
+        super.init(viewMode: .sourceAccurate)
+    }
+
+    /// Visits a node to collect its type name and store it as an ancestor type name to prepend to any
+    /// children to form their fully-qualified type names
+    private func visitNode(_ node: some TypeNameCollectible) -> SyntaxVisitorContinueKind {
+        let name = node.typeName
+        let fullyQualifiedName = (ancestorNames + [name]).joined(separator: ".")
+        names.insert(fullyQualifiedName)
+
+        if !requireFullyQualifiedNames {
+            names.insert(name)
+        }
+
+        ancestorNames.push(node.typeName)
```

Comment:
```suggestion
        ancestorNames.push(name)
```

Reply:
Good catch. This code goes away with the change to remove the protocol.